src/timeline/Timeline.js

```diff
@@ -0,0 +1,208 @@
+// @flow
+import {
+  View,
+  Text,
+  ScrollView,
+  TouchableOpacity,
+  Dimensions
+} from 'react-native';
+import PropTypes from 'prop-types';
+import populateEvents from './Packer';
+import React from 'react';
+import moment from 'moment';
+import _ from 'lodash';
+import styleConstructor from './style';
+
+const LEFT_MARGIN = 60 - 1;
+// const RIGHT_MARGIN = 10
+// const CALENDER_HEIGHT = 2400;
+// const EVENT_TITLE_HEIGHT = 15
+const TEXT_LINE_HEIGHT = 17;
+// const MIN_EVENT_TITLE_WIDTH = 20
+// const EVENT_PADDING_LEFT = 4
+
+function range(from, to) {
+  return Array.from(Array(to), (_, i) => from + i);
+}
+
+let {width: dimensionWidth} = Dimensions.get('window');
+
+export default class Timeline extends React.PureComponent {
+  static propTypes = {
+    start: PropTypes.number,
+    end: PropTypes.number,
+    eventTapped: PropTypes.func,
+    format24h: PropTypes.bool,
+    events: PropTypes.arrayOf(PropTypes.shape({
+      start: PropTypes.string.isRequired,
+      end: PropTypes.string.isRequired,
+      title: PropTypes.string.isRequired,
+      summary: PropTypes.string.isRequired,
+      color: PropTypes.string
+    })).isRequired
+  }
+
+  static defaultProps = {
+    start: 0,
+    end: 24,
+    events: [],
+    format24h: true
+  }
+
+  constructor(props) {
+    super(props);
+    const {start, end} = this.props;
+    this.calendarHeight = (end - start) * 100;
+    this.styles = styleConstructor(props.styles, this.calendarHeight);
+    const width = dimensionWidth - LEFT_MARGIN;
+    const packedEvents = populateEvents(props.events, width, start);
+    let initPosition =
+      _.min(_.map(packedEvents, 'top')) - this.calendarHeight / (end - start);
+    const verifiedInitPosition = initPosition < 0 ? 0 : initPosition;
+    this.state = {
+      _scrollY: verifiedInitPosition,
+      packedEvents
+    };
+  }
+
+  UNSAFE_componentWillReceiveProps({events, start = 0}) {
+    const width = dimensionWidth - LEFT_MARGIN;
+    this.setState({
+      packedEvents: populateEvents(events, width, start)
+    });
+  }
```

Comment:
Can we please find another lifecycle which isn't UNSAFE ? :)

Reply:
Fixed All of Them.